crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,1006 @@
+        }
+        nodes
+    }
+}
+
+impl<'a, T, V, Ix> Iterator for Iter<'a, T, V, Ix>
+where
+    Ix: IndexType,
+{
+    type Item = (&'a Interval<T>, &'a V);
+
+    #[allow(clippy::unwrap_used, clippy::unwrap_in_result)]
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_none() {
+            self.init_stack();
+        }
+        let stack = self.stack.as_mut().unwrap();
+        if stack.is_empty() {
+            return None;
+        }
+        let x = stack.pop().unwrap();
+        stack.extend(Self::left_tree(
+            self.map_ref,
+            self.map_ref.nref(x, Node::right),
+        ));
+        Some(self.map_ref.nref(x, |xn| (xn.interval(), xn.value())))
+    }
+}
+
+/// A view into a single entry in a map, which may either be vacant or occupied.
+#[allow(clippy::exhaustive_enums)] // It is final
+#[derive(Debug)]
+pub enum Entry<'a, T, V, Ix> {
+    /// An occupied entry.
+    Occupied(OccupiedEntry<'a, T, V, Ix>),
+    /// A vacant entry.
+    Vacant(VacantEntry<'a, T, V, Ix>),
+}
+
+/// A view into an occupied entry in a `IntervalMap`.
+/// It is part of the [`Entry`] enum.
+#[derive(Debug)]
+pub struct OccupiedEntry<'a, T, V, Ix> {
+    /// Reference to the map
+    map_ref: &'a mut IntervalMap<T, V, Ix>,
+    /// The entry node
+    node: NodeIndex<Ix>,
+}
+
+/// A view into a vacant entry in a `IntervalMap`.
+/// It is part of the [`Entry`] enum.
+#[derive(Debug)]
+pub struct VacantEntry<'a, T, V, Ix> {
+    /// Mutable reference to the map
+    map_ref: &'a mut IntervalMap<T, V, Ix>,
+    /// The interval of this entry
+    interval: Interval<T>,
+}
+
+impl<'a, T, V, Ix> Entry<'a, T, V, Ix>
+where
+    T: Ord + Clone,
+    Ix: IndexType,
+{
+    /// Ensures a value is in the entry by inserting the default if empty, and returns
+    /// a mutable reference to the value in the entry.
+    #[inline]
+    pub fn or_insert(self, default: V) -> &'a mut V {
+        match self {
+            Entry::Occupied(entry) => entry.map_ref.nmut(entry.node, Node::value_mut),
+            Entry::Vacant(entry) => {
+                let entry_idx = NodeIndex::new(entry.map_ref.nodes.len());
+                let _ignore = entry.map_ref.insert(entry.interval, default);
+                entry.map_ref.nmut(entry_idx, Node::value_mut)
+            }
+        }
+    }
+
+    /// Provides in-place mutable access to an occupied entry before any
+    /// potential inserts into the map.
+    ///
+    /// # Panics
+    ///
+    /// This method panics when the node is a sentinel node
+    #[inline]
+    #[must_use]
+    pub fn and_modify<F>(self, f: F) -> Self
+    where
+        F: FnOnce(&mut V),
+    {
+        match self {
+            Entry::Occupied(entry) => {
+                f(entry.map_ref.nmut(entry.node, Node::value_mut));
+                Self::Occupied(entry)
+            }
+            Entry::Vacant(entry) => Self::Vacant(entry),
+        }
+    }
+}
+
+// TODO: better typed `Node`
```

Comment:
What does better typed mean here? Do we still need it?

Reply:
Yes, we might still need it, I plan to put sentinel node to a seperated type, but it's not required for now.

Reply:
OK，then leave an issue to track it.

Reply:
Tracked in #688